### ISSUE TITLE
[HOLD for payment 2022-02-08] Android - Chat - App crashes when attaching a large size image

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#6836]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Open app and login
2. Open any chat
3. Tap on + > Add Attachment > Choose from galery
4. Select a very long image (use the one attached to the issue)
5. Wait a moment

## Expected Result:
Preview of the  image is displayed and able to send the image

## Actual Result:
The app closes / crashes

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Android

**Version Number:** 1.1.22-0

**Reproducible in staging?:** Yes

**Reproducible in production?:** No

**Logs:** https://stackoverflow.com/c/expensify/questions/4856

**Notes/Photos/Videos:** 
![Bug5380462_longimage2](https://user-images.githubusercontent.com/43996225/146659081-f92b3435-5bc1-4d8f-8416-4904b6b157ff.jpg)

**Expensify/Expensify Issue URL:**

https://user-images.githubusercontent.com/43996225/146659084-e91b859c-3284-4a90-b06b-95669082764b.mp4

**Issue reported by:** Applause

**Slack conversation:** 

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by MelvinBot** [comment]
Triggered auto assignment to @madmax330 (`Engineering`), see https://stackoverflow.com/c/expensify/questions/4319 for more details.

**Comment by sketchydroide** [comment]
I think Max is still ooo, due to time sensitivity, I'm reassigning this GH

**Comment by Jag96** [comment]
Looks like this is still happening on main, I can reproduce on my emulator but there aren't any obvious error messages in the metro logs or Expensify logs, so going to see if I can reproduce on production (1.1.18-3) and narrow it down from there. Also one thing I noticed, if I re-open the app after the crash and try to upload the same image again it works without any crash.

**Comment by Jag96** [comment]
I can still reproduce this on 1.1.18-3 which has been the production version since December 8th:
![image](https://user-images.githubusercontent.com/3981102/146826230-84523a37-c18d-4075-bd72-63bb2398ca46.png)

@kavimuru can you confirm this works on production consistently? Also, what device was this tested on? Getting the device info could help in finding the crash from the list in crashlytics.

One crash in crashlyitcs that is showing up is [this one](https://console.firebase.google.com/u/0/project/expensify-chat/crashlytics/app/android:com.expensify.chat/issues/6a6156c5cd7578c65f572004cd628c5b?time=last-seven-days&type=all&versions=1.1.22-0%20(1001012200)&sessionEventKey=61BE801E00DD00017ED1011F7DABD53C_1621653820793982505) (note - internal link only), and it only started in 1.1.22-0 so going to look into the deploy checklist at potential issues
```
Fatal Exception: java.lang.RuntimeException
Canvas: trying to draw too large(159088608bytes) bitmap.
```

**Comment by Jag96** [comment]
I'm able to also reproduce this on 1.1.17-7 which was the previous production version, so I don't think this is a deploy blocker. Removing the deploy blocker label, adding Daily, and setting External.

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`External`), see https://stackoverflow.com/c/expensify/questions/8582 for more details.

**Comment by jliexpensify** [comment]
Posted!

Internal - https://www.upwork.com/ab/applicants/1473083482261811200/job-details
External - https://www.upwork.com/jobs/~010f941ebf54ffb6c7

**Comment by MelvinBot** [comment]
Triggered auto assignment to @parasharrajat (`Exported`)

**Comment by MelvinBot** [comment]
Current assignee @Jag96 is eligible for the Exported assigner, not assigning anyone new.

**Comment by ahmdshrif** [comment]
## Proposal
 

this a react-native picker known issue 
and the workaround for this is to add  `maxWidth` and `maxHeight`  properties to options. 

so going to `/src/components/AttachmentPicker/index.native.js`
and add them the recommendation value is maxWidth: 1200, maxHeight: 800, but I try with this bigger value and the issue is not happening will start to happen if I increase any of these values. 

another solution is to define quality as 0.7 also works with me but the image does not get a preview, also it will affect the small image quality. 

```diff
const imagePickerOptions = {
    includeBase64: false,
    saveToPhotos: false,
    selectionLimit: 1,
+  maxWidth: 2000,
+  maxHeight: 8000,
};
```



https://user-images.githubusercontent.com/21364901/146987801-f30eff0d-b0d9-443f-bb67-2f70c198caf0.mov

**Comment by parasharrajat** [comment]
Interesting @ahmdshrif. Do you have a link so that I can verify the issue? What is the maximum size that we can use?

**Comment by ahmdshrif** [comment]
@parasharrajat its recommended by some user not liberary so i try to many value woth myself

**Comment by ahmdshrif** [comment]
this issue link on react-native-image-picker
https://github.com/react-native-image-picker/react-native-image-picker/issues/1073

**Comment by Jag96** [comment]
> now you can update any large image and they will get resized to the max width and height we define in the code

Does this affect the way that smaller images show up? Or does it only re-size if the image is larger than one of those values (2000/8000)?

**Comment by parasharrajat** [comment]
I didn't find any information about resizing the image that fixes this issue. That issue talks more about the size of the image file.  How did you reach to that conclusion @ahmdshrif ?

**Comment by ahmdshrif** [comment]
@Jag96  this is the max-width and max-height, it only affects the image more than the threshold we provide. 

@parasharrajat  the conclusion comes from some searches and tests.   the small Image did not have the issue and the big image had the issue so if we resize the big image the issue is solved.
and buting this approach in test it works. 

I know this is a workaround but I think the real solution will be from SDK itself.

**Comment by parasharrajat** [comment]
Ok. Another question would be Will downsizing the image blur it or will it affect the image visually? If the Answer is No, Then I think we can adopt this solution as app crash is a critical issue.

cc: @thienlnam

:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by ahmdshrif** [comment]
@parasharrajat  by SDK I mean the react-native-image-picker SDK. 

I think the big image will have some quality downgrade it's relative to its size since more big images more a quality downgrade. you can see this effect in a video on the proposal. I do the same test with WhatsApp and I find it downgrades the quality with the same quality I get in the video.

**Comment by parasharrajat** [comment]
Ok. Thanks for pointing that out. As far as, we can zoom in on the image and clearly see the information, I am good with that. 

@Jag96  Over to you.

**Comment by K4tsuki** [comment]
In telegram, the image is sends as file/document and let other applications open it properly.

**Comment by MelvinBot** [comment]
📣 @ahmdshrif You have been assigned to this job by @Jag96!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) 📖

**Comment by jliexpensify** [comment]
Hi @ahmdshrif - I'm having some trouble finding you on Upworks, could you link me your profile? Cheers!

**Comment by ahmdshrif** [comment]
hi @jliexpensify  yes I checked my Upwork account it gets suspended for id verification.  I do this step.
it should Reactivate in one business day .

but I can start work on PR until it activates again.

**Comment by ahmdshrif** [comment]
hi @jliexpensify I already submit the proposal to Upwork 
my Upwork profile https://www.upwork.com/freelancers/~0143a705fba636cfd8?s=996364627857502209

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (First week)

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Second week)

**Comment by Julesssss** [comment]
Taking over from Joe.

**Comment by ahmdshrif** [comment]
@Julesssss do you will review the pr or joe will review it ?

**Comment by Julesssss** [comment]
Just reviewed it, left a comment

**Comment by murataka** [comment]
**Proposal**

Hello, I know that this issue is closed , or just about to being closed. But I have some news ...

I saw that ,this happens even using our high-end phones , but  the hardcoded max displayable image size is 4 times bigger than the shared above. So I did a detailed research ( so please read it all ,more than 10 hours gone on my side ) , and will share just the summary here, but please feel free to ask any details .

RecordingCanjas.java // android 30 
`public static final int MAX_BITMAP_SIZE = 100 * 1024 * 1024; // 100 MB` 

Here is the latest release (play store version) of the app's logs ( A quick look )  : 
<img width="1355" alt="Screen Shot 2022-01-30 at 06 28 21" src="https://user-images.githubusercontent.com/5358438/151685783-fe640777-fe33-4e3e-9b5b-f6c236b1e892.png">



And this one is just before the crash : 


<img width="1346" alt="Screen Shot 2022-01-30 at 06 29 48" src="https://user-images.githubusercontent.com/5358438/151685805-d58c82eb-db56-4f1c-8466-3fd7363acef5.png">


The expected behavior is to use hermes , and jsc is to be used as a fallback. That being said , In release mode , we possibly never could enable hermes . By the way, jsc uses a lot more memory than hermes, and that is why we can not display big images. Also , hermes is known to work faster.

Solution : There are things to do with the build-deploy system . And be sure that proguard does not remove hermes related libraries , and the release should not try to load debug versions of hermes libraries.

You can download the tested app from this link : 

https://play.google.com/store/apps/details?id=com.expensify.chat&hl=en&gl=US

**Comment by parasharrajat** [comment]
@murataka long story short, you are suggesting to use hermes.

**Comment by murataka** [comment]
> @murataka long story short, you are suggesting using Hermes.
> 
> If yes, we already do that.


> The expected behavior is to use hermes , and jsc is to be used as a fallback. That being said , In release mode , we possibly never could enable hermes vm .

I said "could" if you read care fully. 

It already works in debug mode. I checked every single version what is being used, what is not being used . For some reason , it is not working on the release . Here is a similar case : https://github.com/facebook/react-native/issues/26400

This is why I asked you to read carefully. The above error logs clearly tell hermes is there, but not working .

**Comment by murataka** [comment]
@parasharrajat , evaluating this proposal would require following steps : 

1. Select an old branch which not includes the previous fixes 
2. Create bundle in release mode
2. publish in play store for internal testing ( alpha or pre alpha )
3. Run the application and watch the logcat
4. Create a debug build
5. Check if the issue appears on debug build and compare the logcats 

As this may require some more priviledges.  If you don't already have those priviledges to follow these steps, you can still evaluate it by running from the debug build ( npm run android ).

**Comment by murataka** [comment]
I did everything to crash the app, attached, downloaded , zoomed etc. It does not crash. The current solution does not let me pan and zoom to the image when it is large, if I'm not wrong. 

This is the commit I used for testing : ( you can create a fork from this commit for the same version )  https://github.com/Expensify/App/commit/2503fb40cc6a1dfbb6e602cf98cc48e520bf1969

Changes made : 

```
project.ext.react = [
    enableHermes: false,  // clean and rebuild if changing
]
```

 Emulator with 512 mb memory configured. 1024x600 resolution. 


https://user-images.githubusercontent.com/5358438/151718330-0a6c49d0-fe12-4ff9-8c4c-adf0d2260678.mov

**Comment by Julesssss** [comment]
@murataka first, thank you for taking the time to investigate this.

I thought that we weren't using ProGuard currently (I have had the 'Introduce ProGuard' issue on my todo list for a while). But, it would troubling if Heremes is indeed not being used as we expect (even if it is just in some cases). 

I'm going to follow your steps tomorrow morning, and I will let you know if I see the same results.

**Comment by murataka** [comment]
The reason may ( not ) be proguard, but the above steps would help to find the core issue . Configurations are too complicated to investigate directly. I hope at least one of your test builds will not crash and we shall be able to compare with the current release.

**Comment by Julesssss** [comment]
Hi @murataka. I read through all the linked issues and ran some tests today. I get slightly different results, so would you mind clarifying a few things for me?

> I saw that this happens even using our high-end phones , but the hardcoded max displayable image size is bigger than the shared long jpeg above.

- Does this mean that you are seeing the crash when uploading a file **larger** than 100MB? It's not clear to me whether you see this with smaller images too.
- Is this crash handled by the app currently? Is it closing the app for users, or is the crash just seen in LogCat?

I wasn't able to reproduce any crash or see any issues on any build today, but maybe my conditions don't match yours? I tested with a 5.6MB image (the one shown above). and built the following to Pixel 2 & Nexus 5x devices:
- AppStore signed App bundler build
- Local debug build
- Local signed APK build

If we are able to reproduce this then I would be happy to create a new issue, but we'll first need some more information and reproduction steps. Please let me know what you think!

**Comment by murataka** [comment]
> * Does this mean that you are seeing the crash when uploading a file **larger** than 100MB? It's not clear to me whether you see this with smaller images too.

No. I test with same image, with the current release version from play store, because It does not crash with the above image in debug mode , even before the workaround that was merged a few days ago. 



> * Is this crash handled by the app currently? Is it closing the app for users, or is the crash just seen in LogCat?

The app on playstore crashes currently. The accepted solution limits large images (downscales) , in fact there are many issues with the latest merge about the accepted solution to this issue. I can add screen recordings in next post if you really need them. Also I shall possibly show it crashing again after the next release .



> I wasn't able to reproduce any crash on any build today, but maybe my conditions don't match yours? I tested with a 5.6MB image (the one shown above). and built the following to Pixel 2 & Nexus 5x devices:

Yes, as the accepted solution is a workaround , and creates other issues, and even may crash for some other sized image , I tested and searched what the proper solution would be by using : 



> This is the commit I used for this : ( you can create a fork from this commit for the same version ) [2503fb4](https://github.com/Expensify/App/commit/2503fb40cc6a1dfbb6e602cf98cc48e520bf1969)
> 
> Changes made :
> 
> ```
> project.ext.react = [
>     enableHermes: false,  // clean and rebuild if changing
> ]
> ```
> 
> Emulator with 512 mb memory configured. 1024x600 resolution.

**Comment by murataka** [comment]
I. edited the steps, please be sure about you completely understand why I'm further investigating this before spending time for creating those test builds, as all these are taking additional time , and it is not easy to test before creating the builds.

**Comment by Julesssss** [comment]
I understand why you are investigating.

> No. I test with same image, with the current release version from play store

Okay. So I did the same thing as you, but I am not seeing a crash. I am on the app version`1.1.33-3` maybe newer than yours as I am on the beta channel. What version are you testing, please?

> there are many issues with the latest merge about the accepted solution to this issue. I can add screen recordings in next post if you really need them

Reproducible steps and information about the exact device and app version you are seeing the crash on is preferable at this point. I don't want to create a new issue for this before someone else has been able to reproduce.

**Comment by murataka** [comment]
1.1.32-0 is what I can download in release mode . 

> I don't want to create a new issue for this before someone else has been able to reproduce

There's no new issue , I suspect the actual one is not solved already .  



> So I did the same thing as you, but I am not seeing a crash. I am on the app version`1.1.33-3` maybe newer than yours

When you test the above fix ( workaround ) and the same image , you should it would not crash.

This was the accepted proposal , the reviewers may not think this is an issue , but I think there are many issues with that. 

I saw there are many opened issues about much smaller bugs than that workaround creates. But again , if maintainers think they don't think they need a proper fix, and the workaround is fine , we should not spend more time on this . However , when I shall have time , I shall share some video recordings about that workaround creates.

As being said, solving those issues will require that workaround to be deleted , and the actual fix to be implemented .

> ## Proposal
> this a react-native picker known issue and the workaround for this is to add `maxWidth` and `maxHeight` properties to options.
> 
> so going to `/src/components/AttachmentPicker/index.native.js` and add them the recommendation value is maxWidth: 1200, maxHeight: 800, but I try with this bigger value and the issue is not happening will start to happen if I increase any of these values.
> 
> another solution is to define quality as 0.7 also works with me but the image does not get a preview, also it will affect the small image quality.

**Comment by Julesssss** [comment]
> As being said, solving those issues will require that workaround to be deleted , and the actual fix to be implemented .

I understand that. 

> There's no new issue

By issue, I mean a Github issue. The previous fix was acceptable at the time, so I don't think this should be a regression, but a new followup issue.

You have made some interesting points and if we are able to come up with a proper proposal I would like for us to create a new issue, else we will not be able to pay you for your work 😄 

As far as we are concerned, this previous solution was okay. Please forget about that for now. I would like you to approach the proposal as a new problem to solve -- an improvement to the previous solution. Does that make sense?

**Comment by Julesssss** [comment]
> I hope someone will not report the same bug with a different title with 250 , and someone else solve the actual issue for 250 again at the mean time.

This will not happen. It is clearly your proposal, you have the comments above to prove it.

> Yes , but the improvement will require deletion of the actual fix

If PRs could only add code and not remove existing code it would be a very different world 😆 

No rush! Please do not start on this until we have 1) clarified what your exact proposal is, 2) Created a new issue and hired you.

**Comment by mvtglobally** [comment]
Issue not reproducible during KI retests. (Third week)
@Julesssss  Do we need to adjust any steps to repro this? I see you discuss that issue is persisting.

**Comment by Julesssss** [comment]
Hi @mvtglobally, please ignore this for now.

**Comment by botify** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.1.33-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/7209


If no regressions arise, payment will be issued on 2022-02-08. :confetti_ball:

**Comment by murataka** [comment]
![IMG_20220202_124338](https://user-images.githubusercontent.com/5358438/152130082-bd012fcf-db5d-4bc0-b533-490ad46232c6.jpg)

**Comment by murataka** [comment]
![rotated](https://user-images.githubusercontent.com/5358438/152130446-dbcfe7f7-da78-4f17-b422-38a17f89def8.jpg)

**Comment by ahmdshrif** [comment]
I want to mention 2 points here: - 

1- in my pr, I am not resizing the width of the image. they have already resized it to be the width of the screen.
2- we have some discussion on pr on other solutions 
if we disable the hardware acceleration this issue will not happen and we will not need to resize. (but that suggestion was rejected because we need hardware acceleration)

I just mention that may be helped in what Murata tries to do but don't get exactly his suggestion.

**Comment by murataka** [comment]
@ahmdshrif , I agree with  your comments, could you open discussion on slack , and maybe we share conclusions here , to keep things simple 😄

**Comment by murataka** [comment]
https://expensify.slack.com/archives/C01GTK53T8Q/p1643796348343759

**Comment by murataka** [comment]
> 2- we have some discussion on pr on other solutions

Is it the right way me to join the pr discussion to help on this ? I'm a bit confused.

**Comment by Julesssss** [comment]
@murataka please can you start again from the beginning of the contribution process. We can continue in the Slack thread for now. Once you have proposed the change we can create an issue and get started. but first, we need more information and more precise information.

Please re-read [these](https://github.com/Expensify/App/blob/main/HOW_TO_CREATE_A_PLAN.md) [docs](https://github.com/Expensify/App/blob/main/CONTRIBUTING.md) too.

**Comment by murataka** [comment]
> @murataka please can you start again from the beginning of the contribution process.

you mean no more proposals here ?

**Comment by jliexpensify** [comment]
Will close once Rajat is paid. Thanks everyone!
